### PR TITLE
Fix bug when `iloc` slice terminates at before-the-zero position

### LINKED ISSUE #7246: [BUG] iloc not working properly
**Describe the bug**
It seems that specific uses of `iloc` on a cudf dataframe do not return the correct results (or anything). 

**Steps/Code to reproduce bug**
```
import cudf
import pandas as pd
data = [0,1,2,3,4]

df = pd.DataFrame(data)
df.iloc[::-1] # returns correct result on CPU of [4,3,2,1,0]

df = cudf.DataFrame(data)
df.iloc[::-1] # returns empty dataframe 
```

**Expected behavior**
I would expect the CPU and GPU result to be identical.

**Environment overview (please complete the following information)**
 - Environment location: DGX Station
 - Method of cuDF install: Docker
   - `docker pull rapidsai/rapidsai:cuda10.1-runtime-ubuntu18.04`

**Environment details**

<details><summary>Click here to see environment details</summary><pre>
     
     **git***
     commit ecd508c4d332ea0866a9a004ae4294f0b1445f59 (HEAD -> branch-0.18, origin/branch-0.18, origin/HEAD)
     Author: Patrick Liu <34781700+ptartan21@users.noreply.github.com>
     Date:   Wed Jan 13 12:07:26 2021 -0600
     
     Exposing model_selection in a similar way to scikit-learn(#3329)
     
     Resolving #3267. It seems that model_selection is already properly exposed through `cuml.preprocessing.model_selection` A small test suite for `train_test_split` is included in this PR to demonstrate that it works as desired.
     
     Authors:
     - Patrick Liu <patrick.liu01@gmail.com>
     
     Approvers:
     - William Hicks (@wphicks)
     - John Zedlewski (@JohnZed)
     
     URL: https://github.com/rapidsai/cuml/pull/3329
     **git submodules***
     
     ***OS Information***
     DISTRIB_ID=Ubuntu
     DISTRIB_RELEASE=18.04
     DISTRIB_CODENAME=bionic
     DISTRIB_DESCRIPTION="Ubuntu 18.04.5 LTS"
     NAME="Ubuntu"
     VERSION="18.04.5 LTS (Bionic Beaver)"
     ID=ubuntu
     ID_LIKE=debian
     PRETTY_NAME="Ubuntu 18.04.5 LTS"
     VERSION_ID="18.04"
     HOME_URL="https://www.ubuntu.com/"
     SUPPORT_URL="https://help.ubuntu.com/"
     BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
     PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
     VERSION_CODENAME=bionic
     UBUNTU_CODENAME=bionic
     Linux c561bb98dbf9 4.15.0-134-generic #138-Ubuntu SMP Fri Jan 15 10:52:18 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
     
     ***GPU Information***
     Thu Jan 28 19:23:41 2021
     +-----------------------------------------------------------------------------+
     | NVIDIA-SMI 418.181.07   Driver Version: 418.181.07   CUDA Version: 10.1     |
     |-------------------------------+----------------------+----------------------+
     | GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
     | Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
     |===============================+======================+======================|
     |   0  Tesla V100-DGXS...  On   | 00000000:07:00.0  On |                    0 |
     | N/A   32C    P0    50W / 300W |   3298MiB / 16125MiB |      0%      Default |
     +-------------------------------+----------------------+----------------------+
     |   1  Tesla V100-DGXS...  On   | 00000000:08:00.0 Off |                    0 |
     | N/A   32C    P0    36W / 300W |     12MiB / 16128MiB |      0%      Default |
     +-------------------------------+----------------------+----------------------+
     |   2  Tesla V100-DGXS...  On   | 00000000:0E:00.0 Off |                    0 |
     | N/A   32C    P0    36W / 300W |     12MiB / 16128MiB |      0%      Default |
     +-------------------------------+----------------------+----------------------+
     |   3  Tesla V100-DGXS...  On   | 00000000:0F:00.0 Off |                    0 |
     | N/A   33C    P0    36W / 300W |     12MiB / 16128MiB |      0%      Default |
     +-------------------------------+----------------------+----------------------+
     
     +-----------------------------------------------------------------------------+
     | Processes:                                                       GPU Memory |
     |  GPU       PID   Type   Process name                             Usage      |
     |=============================================================================|
     +-----------------------------------------------------------------------------+
     
     ***CPU***
     Architecture:        x86_64
     CPU op-mode(s):      32-bit, 64-bit
     Byte Order:          Little Endian
     CPU(s):              40
     On-line CPU(s) list: 0-39
     Thread(s) per core:  2
     Core(s) per socket:  20
     Socket(s):           1
     NUMA node(s):        1
     Vendor ID:           GenuineIntel
     CPU family:          6
     Model:               79
     Model name:          Intel(R) Xeon(R) CPU E5-2698 v4 @ 2.20GHz
     Stepping:            1
     CPU MHz:             1199.268
     CPU max MHz:         3600.0000
     CPU min MHz:         1200.0000
     BogoMIPS:            4397.40
     Virtualization:      VT-x
     L1d cache:           32K
     L1i cache:           32K
     L2 cache:            256K
     L3 cache:            51200K
     NUMA node0 CPU(s):   0-39
     Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cdp_l3 invpcid_single pti intel_ppin ssbd ibrs ibpb stibp tpr_shadow vnmi flexpriority ept vpid fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm cqm rdt_a rdseed adx smap intel_pt xsaveopt cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local dtherm ida arat pln pts md_clear flush_l1d
     
     ***CMake***
     
     ***g++***
     
     ***nvcc***
     
     ***Python***
     /opt/conda/envs/rapids/bin/python
     Python 3.7.8
     
     ***Environment Variables***
     PATH                            : /opt/conda/envs/rapids/bin:/opt/conda/condabin:/opt/conda/bin:/usr/local/nvidia/bin:/usr/local/cuda/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
     LD_LIBRARY_PATH                 : /opt/conda/envs/rapids/lib:/usr/local/nvidia/lib:/usr/local/nvidia/lib64
     NUMBAPRO_NVVM                   :
     NUMBAPRO_LIBDEVICE              :
     CONDA_PREFIX                    : /opt/conda/envs/rapids
     PYTHON_PATH                     :
     
     ***conda packages***
     /opt/conda/condabin/conda
     # packages in environment at /opt/conda/envs/rapids:
     #
     # Name                    Version                   Build  Channel
     _libgcc_mutex             0.1                 conda_forge    conda-forge
     _openmp_mutex             4.5                      1_llvm    conda-forge
     abseil-cpp                20200225.2           he1b5a44_2    conda-forge
     aiohttp                   3.7.3            py37h4abf009_0    conda-forge
     alsa-lib                  1.2.3                h516909a_0    conda-forge
     appdirs                   1.4.4              pyh9f0ad1d_0    conda-forge
     argon2-cffi               20.1.0           py37h4abf009_2    conda-forge
     arrow-cpp                 1.0.1           py37h2318771_14_cuda    conda-forge
     arrow-cpp-proc            2.0.0                      cuda    conda-forge
     async-timeout             3.0.1                   py_1000    conda-forge
     async_generator           1.10                       py_0    conda-forge
     attrs                     20.3.0             pyhd3deb0d_0    conda-forge
     aws-c-common              0.4.59               h36c2ea0_1    conda-forge
     aws-c-event-stream        0.1.6                had2084c_6    conda-forge
     aws-checksums             0.1.10               h4e93380_0    conda-forge
     aws-sdk-cpp               1.8.63               h9b98462_0    conda-forge
     backcall                  0.2.0              pyh9f0ad1d_0    conda-forge
     backports                 1.0                        py_2    conda-forge
     backports.functools_lru_cache 1.6.1                      py_0    conda-forge
     blas                      2.17                   openblas    conda-forge
     blazingsql                0.17.0a0                 pypi_0    pypi
     bleach                    3.2.1              pyh9f0ad1d_0    conda-forge
     bokeh                     2.2.3            py37h89c1867_0    conda-forge
     boost                     1.72.0           py37h48f8a5e_1    conda-forge
     boost-cpp                 1.72.0               h8e57a91_0    conda-forge
     boto3                     1.16.34            pyhd8ed1ab_0    conda-forge
     botocore                  1.19.34            pyhd8ed1ab_0    conda-forge
     brotli                    1.0.9                he1b5a44_3    conda-forge
     brotlipy                  0.7.0           py37hb5d75c8_1001    conda-forge
     bzip2                     1.0.8                h516909a_3    conda-forge
     c-ares                    1.17.1               h36c2ea0_0    conda-forge
     ca-certificates           2020.12.5            ha878542_0    conda-forge
     cairo                     1.16.0            hcf35c78_1003    conda-forge
     certifi                   2020.12.5        py37h89c1867_0    conda-forge
     cffi                      1.14.4           py37h11fe52a_0    conda-forge
     cfitsio                   3.470                h2e3daa1_7    conda-forge
     chardet                   3.0.4           py37he5f6b98_1008    conda-forge
     click                     7.1.2              pyh9f0ad1d_0    conda-forge
     click-plugins             1.1.1                      py_0    conda-forge
     cligj                     0.7.1              pyhd8ed1ab_0    conda-forge
     cloudpickle               1.6.0                      py_0    conda-forge
     colorcet                  2.0.1                      py_0    conda-forge
     cryptography              3.2.1            py37hc72a4ac_0    conda-forge
     cudatoolkit               10.1.243             h6bb024c_0    nvidia
     cudf                      0.17.0a201209   cuda_10.1_py37_gbd321d1e93_382    rapidsai-nightly
     cudf_kafka                0.17.0a201209   py37_gbd321d1e93_382    rapidsai-nightly
     cudnn                     7.6.0                cuda10.1_0    nvidia
     cugraph                   0.17.0a201210   py37_ge205fd07_289    rapidsai-nightly
     cuml                      0.17.0a201210   cuda10.1_py37_g2c0aacf44_173    rapidsai-nightly
     cupy                      8.0.0            py37h0632833_0    conda-forge
     curl                      7.71.1               he644dc0_8    conda-forge
     cusignal                  0.17.0a201210   py37_g44e57f5_47    rapidsai-nightly
     cuspatial                 0.17.0a201210   py37_g9aeebc8_23    rapidsai-nightly
     custreamz                 0.17.0a201209   py37_gbd321d1e93_382    rapidsai-nightly
     cuxfilter                 0.17.0a201203   py37_g01ee2c8_96    rapidsai-nightly
     cycler                    0.10.0                     py_2    conda-forge
     cyrus-sasl                2.1.27               h063b49f_1    conda-forge
     cython                    0.29.21          py37hb892b2f_1    conda-forge
     cytoolz                   0.11.0           py37h4abf009_1    conda-forge
     dask                      2.30.0                     py_0    conda-forge
     dask-core                 2.30.0                     py_0    conda-forge
     dask-cuda                 0.17.0a201207           py37_56    rapidsai-nightly
     dask-cudf                 0.17.0a201209   py37_gbd321d1e93_382    rapidsai-nightly
     dask-glm                  0.2.0                      py_1    conda-forge
     dask-labextension         3.0.0                      py_0    conda-forge
     dask-ml                   1.7.0                      py_0    conda-forge
     datashader                0.11.1             pyh9f0ad1d_0    conda-forge
     datashape                 0.5.4                      py_1    conda-forge
     decorator                 4.4.2                      py_0    conda-forge
     defusedxml                0.6.0                      py_0    conda-forge
     distributed               2.30.1           py37h89c1867_0    conda-forge
     dlpack                    0.3                  he1b5a44_1    conda-forge
     entrypoints               0.3             pyhd8ed1ab_1003    conda-forge
     expat                     2.2.9                he1b5a44_2    conda-forge
     fa2                       0.3.5            py37h8f50634_0    conda-forge
     faiss-proc                1.0.0                      cuda    rapidsai-nightly
     fastavro                  1.2.0            py37h4abf009_0    conda-forge
     fastrlock                 0.5              py37h3340039_1    conda-forge
     filterpy                  1.4.5                      py_1    conda-forge
     fiona                     1.8.13           py37h0492a4a_1    conda-forge
     fontconfig                2.13.1            h86ecdb6_1001    conda-forge
     freetype                  2.10.4               h7ca028e_0    conda-forge
     freexl                    1.0.5             h516909a_1002    conda-forge
     fsspec                    0.8.4                      py_0    conda-forge
     future                    0.18.2           py37h89c1867_2    conda-forge
     gdal                      3.0.4           py37h4b180d9_10    conda-forge
     geopandas                 0.8.1                      py_0    conda-forge
     geos                      3.8.1                he1b5a44_0    conda-forge
     geotiff                   1.6.0                h05acad5_0    conda-forge
     gettext                   0.19.8.1          hf34092f_1004    conda-forge
     gflags                    2.2.2             he1b5a44_1004    conda-forge
     giflib                    5.2.1                h36c2ea0_2    conda-forge
     git                       2.29.2          pl5320h78be0e9_1    conda-forge
     glib                      2.66.3               h58526e2_0    conda-forge
     glog                      0.4.0                h49b9bf7_3    conda-forge
     google-cloud-cpp          1.16.0               he4a878c_2    conda-forge
     google-cloud-cpp-common   0.25.0               he83eced_7    conda-forge
     googleapis-cpp            0.10.0               h6b1abdc_4    conda-forge
     gpuci-tools               0.2.1                         5    gpuci
     graphite2                 1.3.13            h58526e2_1001    conda-forge
     grpc-cpp                  1.32.0               h7997a97_1    conda-forge
     gtest                     1.10.0               h0efe328_4    conda-forge
     harfbuzz                  2.4.0                h9f30f68_3    conda-forge
     hdf4                      4.2.13            hf30be14_1003    conda-forge
     hdf5                      1.10.6          nompi_h7c3c948_1111    conda-forge
     heapdict                  1.0.1                      py_0    conda-forge
     holoviews                 1.14.0             pyhd3deb0d_0    conda-forge
     icu                       64.2                 he1b5a44_1    conda-forge
     idna                      2.10               pyh9f0ad1d_0    conda-forge
     importlib-metadata        3.1.1              pyhd8ed1ab_0    conda-forge
     importlib_metadata        3.1.1                hd8ed1ab_0    conda-forge
     iniconfig                 1.1.1              pyh9f0ad1d_0    conda-forge
     ipykernel                 5.4.0            py37h888b3d9_0    conda-forge
     ipython                   7.15.0           py37hc8dfbb8_0    conda-forge
     ipython_genutils          0.2.0                      py_1    conda-forge
     ipywidgets                7.5.1              pyh9f0ad1d_1    conda-forge
     jedi                      0.17.2           py37h89c1867_1    conda-forge
     jinja2                    2.11.2             pyh9f0ad1d_0    conda-forge
     jmespath                  0.10.0             pyh9f0ad1d_0    conda-forge
     joblib                    0.17.0                     py_0    conda-forge
     jpeg                      9d                   h36c2ea0_0    conda-forge
     jpype1                    1.1.2            py37h99015e2_0    conda-forge
     json-c                    0.13.1            hbfbb72e_1002    conda-forge
     json5                     0.9.5              pyh9f0ad1d_0    conda-forge
     jsonschema                3.2.0                      py_2    conda-forge
     jupyter-server-proxy      1.5.0                      py_0    conda-forge
     jupyter_client            6.1.7                      py_0    conda-forge
     jupyter_core              4.7.0            py37h89c1867_0    conda-forge
     jupyterlab                2.1.5                      py_0    conda-forge
     jupyterlab-nvdashboard    0.1.11200212              py_12    rapidsai-nightly
     jupyterlab_pygments       0.1.2              pyh9f0ad1d_0    conda-forge
     jupyterlab_server         1.2.0                      py_0    conda-forge
     kealib                    1.4.13               h33137a7_1    conda-forge
     kiwisolver                1.3.1            py37hc928c03_0    conda-forge
     krb5                      1.17.2               h926e7f8_0    conda-forge
     lcms2                     2.11                 hcbb858e_1    conda-forge
     ld_impl_linux-64          2.35.1               hed1e6ac_0    conda-forge
     libblas                   3.8.0               17_openblas    conda-forge
     libcblas                  3.8.0               17_openblas    conda-forge
     libcrc32c                 1.1.1                he1b5a44_2    conda-forge
     libcudf                   0.17.0a201210   cuda10.1_g00ca24625e_383    rapidsai-nightly
     libcudf_kafka             0.17.0a201209   gbd321d1e93_382    rapidsai-nightly
     libcugraph                0.17.0a201210   cuda10.1_ge205fd07_289    rapidsai-nightly
     libcuml                   0.17.0a201210   cuda10.1_g2c0aacf44_173    rapidsai-nightly
     libcumlprims              0.17.0a201030   cuda10.1_g1fa28a5_8    rapidsai-nightly
     libcurl                   7.71.1               hcdd3856_8    conda-forge
     libcuspatial              0.17.0a201210   cuda10.1_g9aeebc8_23    rapidsai-nightly
     libdap4                   3.20.6               h1d1bd15_1    conda-forge
     libedit                   3.1.20191231         he28a2e2_2    conda-forge
     libev                     4.33                 h516909a_1    conda-forge
     libevent                  2.1.10               hcdb4288_3    conda-forge
     libfaiss                  1.6.3           he68dc02_3_cuda    conda-forge
     libffi                    3.2.1             he1b5a44_1007    conda-forge
     libgcc-ng                 7.5.0               h5dbcf3e_17    conda-forge
     libgcrypt                 1.8.7                h36c2ea0_0    conda-forge
     libgdal                   3.0.4               he6a97d6_10    conda-forge
     libgfortran-ng            7.5.0               hae1eefd_17    conda-forge
     libgfortran4              7.5.0               hae1eefd_17    conda-forge
     libglib                   2.66.3               hbe7bbb4_0    conda-forge
     libgpg-error              1.39                 he1b5a44_0    conda-forge
     libgsasl                  1.8.0                         2    conda-forge
     libhwloc                  2.3.0                h5e5b7d1_1    conda-forge
     libiconv                  1.16                 h516909a_0    conda-forge
     libkml                    1.3.0             hd79254b_1012    conda-forge
     liblapack                 3.8.0               17_openblas    conda-forge
     liblapacke                3.8.0               17_openblas    conda-forge
     libllvm10                 10.0.1               he513fc3_3    conda-forge
     libnetcdf                 4.7.4           nompi_hefab0ff_106    conda-forge
     libnghttp2                1.41.0               h8cfc5f6_2    conda-forge
     libntlm                   1.4               h516909a_1002    conda-forge
     libopenblas               0.3.10          pthreads_hb3c22a3_5    conda-forge
     libpng                    1.6.37               h21135ba_2    conda-forge
     libpq                     12.3                 h255efa7_3    conda-forge
     libprotobuf               3.13.0.1             h8b12597_0    conda-forge
     librdkafka                1.5.2                h40bdf00_0    conda-forge
     librmm                    0.17.0a201210   cuda10.1_gb8c8310_60    rapidsai-nightly
     libsodium                 1.0.18               h36c2ea0_1    conda-forge
     libspatialindex           1.9.3                he1b5a44_3    conda-forge
     libspatialite             4.3.0a            h2482549_1038    conda-forge
     libssh2                   1.9.0                hab1572f_5    conda-forge
     libstdcxx-ng              7.5.0               h2ae2ef3_17    conda-forge
     libthrift                 0.13.0               h5aa387f_6    conda-forge
     libtiff                   4.1.0                h4f3a223_6    conda-forge
     libutf8proc               2.6.0                h36c2ea0_0    conda-forge
     libuuid                   2.32.1            h14c3975_1000    conda-forge
     libuv                     1.40.0               hd18ef5c_0    conda-forge
     libwebp                   1.1.0                h76fa15c_4    conda-forge
     libwebp-base              1.1.0                h36c2ea0_3    conda-forge
     libxcb                    1.13              h14c3975_1002    conda-forge
     libxgboost                1.3.0dev.rapidsai0.17      cuda10.1_0    rapidsai-nightly
     libxml2                   2.9.10               hee79883_0    conda-forge
     llvm-openmp               11.0.0               hfc4b9b4_1    conda-forge
     llvmlite                  0.34.0           py37h5202443_2    conda-forge
     locket                    0.2.0                      py_2    conda-forge
     lz4-c                     1.9.2                he1b5a44_3    conda-forge
     markdown                  3.3.3              pyh9f0ad1d_0    conda-forge
     markupsafe                1.1.1            py37hb5d75c8_2    conda-forge
     matplotlib-base           3.3.3            py37h4f6019d_0    conda-forge
     mistune                   0.8.4           py37h4abf009_1002    conda-forge
     more-itertools            8.6.0              pyhd8ed1ab_0    conda-forge
     msgpack-python            1.0.0            py37hc928c03_2    conda-forge
     multidict                 4.7.5            py37h8f50634_2    conda-forge
     multipledispatch          0.6.0                      py_0    conda-forge
     munch                     2.5.0                      py_0    conda-forge
     nbclient                  0.5.1                      py_0    conda-forge
     nbconvert                 6.0.7            py37h89c1867_3    conda-forge
     nbformat                  5.0.8                      py_0    conda-forge
     nccl                      2.8.3.1              h8b44402_0    conda-forge
     ncurses                   6.2                  h58526e2_4    conda-forge
     nest-asyncio              1.4.3              pyhd8ed1ab_0    conda-forge
     netifaces                 0.10.9          py37h8f50634_1003    conda-forge
     networkx                  2.5                        py_0    conda-forge
     nodejs                    12.4.0               he1b5a44_0    conda-forge
     notebook                  6.1.5            py37h89c1867_0    conda-forge
     numba                     0.51.2           py37h9fdb41a_0    conda-forge
     numpy                     1.19.4           py37h7e9df27_1    conda-forge
     nvtx                      0.2.1            py37h8f50634_2    conda-forge
     olefile                   0.46               pyh9f0ad1d_1    conda-forge
     openjdk                   11.0.8               hacce0ff_0    conda-forge
     openjpeg                  2.3.1                h981e76c_3    conda-forge
     openssl                   1.1.1h               h516909a_0    conda-forge
     orc                       1.6.5                hd3605a7_0    conda-forge
     packaging                 20.7               pyhd3deb0d_0    conda-forge
     pandas                    1.1.4            py37h10a2094_0    conda-forge
     pandoc                    2.11.2               h36c2ea0_0    conda-forge
     pandocfilters             1.4.2                      py_1    conda-forge
     panel                     0.10.2             pyhd8ed1ab_0    conda-forge
     param                     1.10.0                     py_0    conda-forge
     parquet-cpp               1.5.1                         2    conda-forge
     parso                     0.7.1              pyh9f0ad1d_0    conda-forge
     partd                     1.1.0                      py_0    conda-forge
     patsy                     0.5.1                      py_0    conda-forge
     pcre                      8.44                 he1b5a44_0    conda-forge
     perl                      5.32.0               h36c2ea0_0    conda-forge
     pexpect                   4.8.0              pyh9f0ad1d_2    conda-forge
     pickle5                   0.0.11           py37h8f50634_0    conda-forge
     pickleshare               0.7.5                   py_1003    conda-forge
     pillow                    8.0.1            py37h63a5d19_0    conda-forge
     pip                       20.3.1             pyhd8ed1ab_0    conda-forge
     pixman                    0.38.0            h516909a_1003    conda-forge
     pluggy                    0.13.1           py37he5f6b98_3    conda-forge
     poppler                   0.87.0               h4190859_1    conda-forge
     poppler-data              0.4.10                        0    conda-forge
     postgresql                12.3                 hc2f5b80_3    conda-forge
     proj                      7.0.0                h966b41f_5    conda-forge
     prometheus_client         0.9.0              pyhd3deb0d_0    conda-forge
     prompt-toolkit            3.0.8              pyha770c72_0    conda-forge
     protobuf                  3.13.0.1         py37h745909e_1    conda-forge
     psutil                    5.7.3            py37hb5d75c8_0    conda-forge
     pthread-stubs             0.4               h36c2ea0_1001    conda-forge
     ptyprocess                0.6.0                   py_1001    conda-forge
     py                        1.9.0              pyh9f0ad1d_0    conda-forge
     py-xgboost                1.3.0dev.rapidsai0.17  cuda10.1py37_0    rapidsai-nightly
     pyarrow                   1.0.1           py37hbeecfa9_14_cuda    conda-forge
     pycparser                 2.20               pyh9f0ad1d_2    conda-forge
     pyct                      0.4.6                      py_0    conda-forge
     pyct-core                 0.4.6                      py_0    conda-forge
     pydeck                    0.5.0              pyh9f0ad1d_0    conda-forge
     pyee                      7.0.4              pyh9f0ad1d_0    conda-forge
     pygments                  2.7.3              pyhd8ed1ab_0    conda-forge
     pyhive                    0.6.2              pyh9f0ad1d_0    conda-forge
     pynvml                    8.0.4                      py_1    conda-forge
     pyopenssl                 20.0.0             pyhd8ed1ab_0    conda-forge
     pyparsing                 2.4.7              pyh9f0ad1d_0    conda-forge
     pyppeteer                 0.2.2                      py_1    conda-forge
     pyproj                    2.6.1.post1      py37h34dd122_0    conda-forge
     pyrsistent                0.17.3           py37h4abf009_1    conda-forge
     pysocks                   1.7.1            py37he5f6b98_2    conda-forge
     pytest                    6.1.2            py37h89c1867_0    conda-forge
     python                    3.7.8           h8bdb77d_2_cpython    conda-forge
     python-confluent-kafka    1.5.0            py37h8f50634_0    conda-forge
     python-dateutil           2.8.1                      py_0    conda-forge
     python_abi                3.7                     1_cp37m    conda-forge
     pytz                      2020.4             pyhd8ed1ab_0    conda-forge
     pyviz_comms               0.7.6              pyh9f0ad1d_0    conda-forge
     pyyaml                    5.3.1            py37hb5d75c8_1    conda-forge
     pyzmq                     20.0.0           py37h5a562af_1    conda-forge
     rapids                    0.17.0a201210   cuda10.1_py37_gdc7a92a_161    rapidsai-nightly
     rapids-blazing            0.17.0a201210   cuda10.1_py37_gdc7a92a_161    rapidsai-nightly
     rapids-xgboost            0.17.0a201210   cuda10.1_py37_gdc7a92a_161    rapidsai-nightly
     re2                       2020.10.01           he1b5a44_0    conda-forge
     readline                  8.0                  he28a2e2_2    conda-forge
     requests                  2.25.0             pyhd3deb0d_0    conda-forge
     rmm                       0.17.0a201210   cuda_10.1_py37_gb8c8310_60    rapidsai-nightly
     rtree                     0.9.4            py37h0b55af0_2    conda-forge
     s3fs                      0.4.2                      py_0    conda-forge
     s3transfer                0.3.3                      py_3    conda-forge
     sasl                      0.2.1           py37h3340039_1002    conda-forge
     scikit-learn              0.23.1           py37h8a51577_0    conda-forge
     scipy                     1.5.1            py37ha3d9a3c_0    conda-forge
     seaborn                   0.11.0               ha770c72_1    conda-forge
     seaborn-base              0.11.0             pyhd8ed1ab_1    conda-forge
     send2trash                1.5.0                      py_0    conda-forge
     setuptools                49.6.0           py37he5f6b98_2    conda-forge
     shapely                   1.7.1            py37hedb1597_1    conda-forge
     simpervisor               0.3                        py_1    conda-forge
     six                       1.15.0             pyh9f0ad1d_0    conda-forge
     snappy                    1.1.8                he1b5a44_3    conda-forge
     sortedcontainers          2.3.0              pyhd8ed1ab_0    conda-forge
     spdlog                    1.7.0                hc9558a2_2    conda-forge
     sqlalchemy                1.3.20           py37h8f50634_0    conda-forge
     sqlite                    3.33.0               h4cf870e_1    conda-forge
     statsmodels               0.12.1           py37ha21ca33_1    conda-forge
     streamz                   0.6.1              pyhd3deb0d_1    conda-forge
     tbb                       2020.2               hc9558a2_0    conda-forge
     tblib                     1.6.0                      py_0    conda-forge
     terminado                 0.9.1            py37h89c1867_1    conda-forge
     testpath                  0.4.4                      py_0    conda-forge
     threadpoolctl             2.1.0              pyh5ca1d4c_0    conda-forge
     thrift                    0.13.0           py37h3340039_2    conda-forge
     thrift_sasl               0.4.2            py37h8f50634_0    conda-forge
     tiledb                    1.7.7                h8efa9f0_3    conda-forge
     tk                        8.6.10               hed695b0_1    conda-forge
     toml                      0.10.2             pyhd8ed1ab_0    conda-forge
     toolz                     0.11.1                     py_0    conda-forge
     tornado                   6.1              py37h4abf009_0    conda-forge
     tqdm                      4.54.1             pyhd8ed1ab_0    conda-forge
     traitlets                 5.0.5                      py_0    conda-forge
     treelite                  0.93             py37h745909e_3    conda-forge
     treelite-runtime          0.93                     pypi_0    pypi
     typing-extensions         3.7.4.3                       0    conda-forge
     typing_extensions         3.7.4.3                    py_0    conda-forge
     tzcode                    2020a                h516909a_0    conda-forge
     ucx                       1.8.1+g6b29558       cuda10.1_0    rapidsai-nightly
     ucx-proc                  1.0.0                       gpu    rapidsai-nightly
     ucx-py                    0.17.0a201210   py37_g6b29558_27    rapidsai-nightly
     umap-learn                0.4.6            py37hc8dfbb8_0    conda-forge
     urllib3                   1.25.11                    py_0    conda-forge
     wcwidth                   0.2.5              pyh9f0ad1d_2    conda-forge
     webencodings              0.5.1                      py_1    conda-forge
     websockets                8.1              py37h8f50634_2    conda-forge
     wheel                     0.36.1             pyhd3deb0d_0    conda-forge
     widgetsnbextension        3.5.1            py37h89c1867_4    conda-forge
     xarray                    0.16.2             pyhd8ed1ab_0    conda-forge
     xerces-c                  3.2.2             h8412b87_1004    conda-forge
     xgboost                   1.3.0dev.rapidsai0.17  cuda10.1py37_0    rapidsai-nightly
     xorg-fixesproto           5.0               h14c3975_1002    conda-forge
     xorg-inputproto           2.3.2             h14c3975_1002    conda-forge
     xorg-kbproto              1.0.7             h14c3975_1002    conda-forge
     xorg-libice               1.0.10               h516909a_0    conda-forge
     xorg-libsm                1.2.3             h84519dc_1000    conda-forge
     xorg-libx11               1.6.12               h516909a_0    conda-forge
     xorg-libxau               1.0.9                h14c3975_0    conda-forge
     xorg-libxdmcp             1.1.3                h516909a_0    conda-forge
     xorg-libxext              1.3.4                h516909a_0    conda-forge
     xorg-libxfixes            5.0.3             h516909a_1004    conda-forge
     xorg-libxi                1.7.10               h516909a_0    conda-forge
     xorg-libxrender           0.9.10            h516909a_1002    conda-forge
     xorg-libxtst              1.2.3             h516909a_1002    conda-forge
     xorg-recordproto          1.14.2            h516909a_1002    conda-forge
     xorg-renderproto          0.11.1            h14c3975_1002    conda-forge
     xorg-xextproto            7.3.0             h14c3975_1002    conda-forge
     xorg-xproto               7.0.31            h14c3975_1007    conda-forge
     xz                        5.2.5                h516909a_1    conda-forge
     yaml                      0.2.5                h516909a_0    conda-forge
     yarl                      1.6.3            py37h4abf009_0    conda-forge
     zeromq                    4.3.3                h58526e2_3    conda-forge
     zict                      2.0.0                      py_0    conda-forge
     zipp                      3.4.0                      py_0    conda-forge
     zlib                      1.2.11            h516909a_1010    conda-forge
     zstd                      1.4.5                h6597ccf_2    conda-forge
     
</pre></details>

**Additional context**
None

### PR DESCRIPTION
**Pull request by isVoid**
Closes #7246

This PR fixes a bug in `Dataframe.iloc`. When the slice provided to `iloc`, is decrementing and also terminates at `before-the-zero` position, such as `slice(2, -1, -1)` or `slice(4, None, -1)`, the terminal position still gets wrapped around. 

`Frame._slice` is moved to `DataFrame._slice` to resolve typing issue.

### CODE REVIEW

**Review by kkraus14 — APPROVED** [review]

**Comment by kkraus14** [comment]
@isVoid can you retarget this to 0.18? This is a nice bugfix with minimal changes that would be good to get in.

**Comment by isVoid** [comment]
FYI I removed typings in the PR. The typing checks revealed some design issues with `Frame._slice`, in that it depends on attributes not exclusively to `Frame`, but to sub-classes. Resolving those issues is beyond the scope of the PR, I'll submit a separate issue for them.

File: python/cudf/cudf/core/dataframe.py [hunk 1]
```diff
@@ -831,6 +833,66 @@ def __sizeof__(self):
         index = self._index.__sizeof__()
         return columns + index
 
+    def _slice(self: T, arg: slice) -> T:
```
Comment (kkraus14) [inline python/cudf/cudf/core/dataframe.py:836]: Does this not get used underneath `Series.iloc[slice(...)]` as well?

Comment (isVoid) [reply python/cudf/cudf/core/dataframe.py:836]: `Series.iloc` seems to depend on `ColumnBase.__getitem__`, which later calls `Series.slice`.

I noticed the following:
```python
>>> s = cudf.Series([1,2,3,4,5])
>>> s.iloc[4:-1:-1]
Series([], dtype: int64)
```
which is likely the same issue, investigating.

Comment (isVoid) [reply python/cudf/cudf/core/dataframe.py:836]: It's the same issue in `ColumnBase.slice`. https://github.com/rapidsai/cudf/blob/b8cb8c72241ca2ac5f4d4064659129e42ddf4dbb/python/cudf/cudf/core/column/column.py#L609-L610

Submitted a quick fix. Although I think the best fix here is to consolidate `Series.iloc` and `DataFrame.iloc` to share the common code path under `Frame._slice`.

**Review by kkraus14 — APPROVED** [review]

**Comment by isVoid** [comment]
rerun tests

**Comment by kkraus14** [comment]
@gpucibot merge

**Comment by isVoid** [comment]
rerun tests

**Comment by galipremsagar** [comment]
@isVoid There seem to be some conflicts, could you resolve them?

**Review by galipremsagar — APPROVED** [review]

**Pull request merged**